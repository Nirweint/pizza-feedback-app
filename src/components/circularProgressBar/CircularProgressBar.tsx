import React, {useEffect} from 'react';

import CircularProgress, {CircularProgressProps} from '@mui/material/CircularProgress';
import Typography from '@mui/material/Typography';
import Box from '@mui/material/Box';

import {ProgressBarType} from "../../types";

export const CircularProgressBar = ({currentProgress, maxValue}: ProgressBarType) => {
  const [progress, setProgress] = React.useState(10);

  useEffect(() => {
    if (maxValue < 1) {
      setProgress(0);
    } else {
      setProgress((100 / maxValue) * currentProgress);
    }
  }, [currentProgress]);

  return <CircularProgressWithLabel value={progress}/>;
};

const progressColorStyle = (progress: number) => {
  if (progress < 40) return 'error';
  if (progress < 70) return 'warning';
  return 'success';
}

const CircularProgressWithLabel = (props: CircularProgressProps & { value: number },) => {
  return (
    <Box sx={{
      position: 'relative',
      display: 'inline-flex',
      width: '100%',
      justifyContent: 'center',
    }}>
      <CircularProgress
        data-testid='circular-progress'
        variant="determinate" {...props}
        size={200}
        color={progressColorStyle(props.value)}
      />
      <Box
        sx={{
          top: 0,
          left: 0,
          bottom: 0,
          right: 0,
          position: 'absolute',
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
        }}
      >
        <Typography
          sx={{fontSize: 25}}
          variant="caption"
          component="div"
          color="text.secondary"
        >{`${Math.round(props.value)}%`}</Typography>
      </Box>
    </Box>
  );
}

