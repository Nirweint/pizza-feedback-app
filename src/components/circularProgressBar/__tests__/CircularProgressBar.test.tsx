import React from 'react';

import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import {CircularProgressBar} from "../CircularProgressBar";

const mockData = {
  maxValue: 10,
  currentProgress: 5,
}

describe('CircularProgressBar', () => {

  const initialState = {};
  const mockStore = configureStore();
  let store;

  test('progress color must be orange', () => {
    store = mockStore(initialState);
    render(<Provider store={store}>
      <CircularProgressBar maxValue={mockData.maxValue}
                           currentProgress={mockData.currentProgress}/>
    </Provider>)

    expect(screen.getByTestId('circular-progress')).toBeInTheDocument();
    expect(screen.getByTestId('circular-progress')).toMatchSnapshot()

  });

  test('progress color must be red', () => {
    store = mockStore(initialState);
    render(<Provider store={store}>
      <CircularProgressBar maxValue={10}
                           currentProgress={2}/>
    </Provider>)

    expect(screen.getByTestId('circular-progress')).toMatchSnapshot()

  });

  test('progress color must be green', () => {
    store = mockStore(initialState);
    render(<Provider store={store}>
      <CircularProgressBar maxValue={10}
                           currentProgress={9}/>
    </Provider>)
    expect(screen.getByTestId('circular-progress')).toMatchSnapshot()

  });

})