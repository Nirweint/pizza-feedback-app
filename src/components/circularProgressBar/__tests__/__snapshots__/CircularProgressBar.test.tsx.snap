// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircularProgressBar progress color must be green 1`] = `
<span
  aria-valuenow="90"
  class="MuiCircularProgress-root MuiCircularProgress-determinate MuiCircularProgress-colorSuccess css-zc2y38-MuiCircularProgress-root"
  data-testid="circular-progress"
  role="progressbar"
  style="width: 200px; height: 200px; transform: rotate(-90deg);"
>
  <svg
    class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
    viewBox="22 22 44 44"
  >
    <circle
      class="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate css-oxts8u-MuiCircularProgress-circle"
      cx="44"
      cy="44"
      fill="none"
      r="20.2"
      stroke-width="3.6"
      style="stroke-dasharray: 126.920; stroke-dashoffset: 12.692px;"
    />
  </svg>
</span>
`;

exports[`CircularProgressBar progress color must be orange 1`] = `
<span
  aria-valuenow="50"
  class="MuiCircularProgress-root MuiCircularProgress-determinate MuiCircularProgress-colorWarning css-osz0mz-MuiCircularProgress-root"
  data-testid="circular-progress"
  role="progressbar"
  style="width: 200px; height: 200px; transform: rotate(-90deg);"
>
  <svg
    class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
    viewBox="22 22 44 44"
  >
    <circle
      class="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate css-oxts8u-MuiCircularProgress-circle"
      cx="44"
      cy="44"
      fill="none"
      r="20.2"
      stroke-width="3.6"
      style="stroke-dasharray: 126.920; stroke-dashoffset: 63.460px;"
    />
  </svg>
</span>
`;

exports[`CircularProgressBar progress color must be red 1`] = `
<span
  aria-valuenow="20"
  class="MuiCircularProgress-root MuiCircularProgress-determinate MuiCircularProgress-colorError css-1baxt8r-MuiCircularProgress-root"
  data-testid="circular-progress"
  role="progressbar"
  style="width: 200px; height: 200px; transform: rotate(-90deg);"
>
  <svg
    class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
    viewBox="22 22 44 44"
  >
    <circle
      class="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate css-oxts8u-MuiCircularProgress-circle"
      cx="44"
      cy="44"
      fill="none"
      r="20.2"
      stroke-width="3.6"
      style="stroke-dasharray: 126.920; stroke-dashoffset: 101.536px;"
    />
  </svg>
</span>
`;
